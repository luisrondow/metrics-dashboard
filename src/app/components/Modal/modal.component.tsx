import { ReactNode, useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { ModalBackground, ModalContent, ModalHeader } from './modal.styles';
import { CloseIcon } from '../icons';

type ModalProps = {
  isOpen: boolean;
  onClose: () => void;
  children: ReactNode;
};

const Modal = ({ isOpen, onClose, children }: ModalProps) => {
  const modalRoot = document.getElementById('modal-root')!;

  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  useEffect(() => {
    if (isOpen) {
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = 'auto';
    }
    return () => {
      document.body.style.overflow = 'auto';
    };
  }, [isOpen]);

  return mounted
    ? createPortal(
        <ModalBackground $isOpen={isOpen} onClick={onClose}>
          <ModalContent data-cy="modal" onClick={(e) => e.stopPropagation()}>
            <ModalHeader data-cy="modal-header">
              <CloseIcon onClick={onClose} />
            </ModalHeader>
            {children}
          </ModalContent>
        </ModalBackground>,
        modalRoot,
      )
    : null;
};

export default Modal;
