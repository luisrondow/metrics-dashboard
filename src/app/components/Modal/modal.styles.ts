import styled, { css } from "styled-components";

type ModalBackgroundProps = {
  $isOpen: boolean;
};

export const ModalBackground = styled.div<ModalBackgroundProps>`
  display: ${({ $isOpen }) => ($isOpen ? 'block' : 'none')};

  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 1000;
`;

export const ModalHeader = styled.div`
  width: 100%;
  display: flex;
  justify-content: flex-end;
`;

export const ModalContent = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background-color: white;
  padding: 20px;
  border-radius: 5px;
  z-index: 1001;
`;