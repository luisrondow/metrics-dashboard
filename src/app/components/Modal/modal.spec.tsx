import { render, screen, fireEvent } from '@testing-library/react';
import Modal from './modal.component';

jest.mock('./modal.styles', () => ({
  ModalBackground: ({
    isOpen,
    children,
    onClick,
  }: {
    isOpen: boolean;
    children: React.ReactNode;
    onClick: () => void;
  }) => (
    <div data-testid="modal-background" style={{ display: isOpen ? 'block' : 'none' }} onClick={onClick}>
      {children}
    </div>
  ),
  ModalContent: ({ children, onClick }: { children: React.ReactNode; onClick: (e: React.MouseEvent) => void }) => (
    <div data-testid="modal-content" onClick={onClick}>
      {children}
    </div>
  ),
  ModalHeader: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
}));

jest.mock('../icons', () => ({
  CloseIcon: ({ onClick }: { onClick: () => void }) => <button onClick={onClick}>Close</button>,
}));

describe('Modal Component', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="modal-root"></div>';
  });

  // add a snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Modal isOpen={true} onClose={() => {}}>
        <div>Modal Content</div>
      </Modal>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders modal when mounted', () => {
    render(
      <Modal isOpen={true} onClose={() => {}}>
        <div>Modal Content</div>
      </Modal>,
    );

    const modalContent = screen.getByText('Modal Content');
    expect(modalContent).toBeInTheDocument();
  });

  it('calls onClose when background is clicked', () => {
    const handleClose = jest.fn();
    render(
      <Modal isOpen={true} onClose={handleClose}>
        <div>Modal Content</div>
      </Modal>,
    );

    const background = screen.getByTestId('modal-background');
    fireEvent.click(background);
    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  it('does not call onClose when content is clicked', () => {
    const handleClose = jest.fn();
    render(
      <Modal isOpen={true} onClose={handleClose}>
        <div>Modal Content</div>
      </Modal>,
    );

    const content = screen.getByTestId('modal-content');
    fireEvent.click(content);
    expect(handleClose).not.toHaveBeenCalled();
  });

  it('calls onClose when close icon is clicked', () => {
    const handleClose = jest.fn();
    render(
      <Modal isOpen={true} onClose={handleClose}>
        <div>Modal Content</div>
      </Modal>,
    );

    const closeButton = screen.getByText('Close');
    fireEvent.click(closeButton);
    expect(handleClose).toHaveBeenCalledTimes(1);
  });
});
