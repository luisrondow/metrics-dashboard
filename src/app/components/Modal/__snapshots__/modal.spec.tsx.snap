// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component matches snapshot 1`] = `<DocumentFragment />`;
